### PR TITLE
Backup/restore with content-defined-chunking (CDC) and new logic

### PR DESCRIPTION
**Pull request by dactoankmapydev**
Backup/restore with content-defined-chunking (CDC)

### CODE REVIEW

**Review by ducpx — APPROVED** [review]

File: pkg/backupapi/file_test.go [hunk 1]
```diff
@@ -1,87 +1,110 @@
...
-
-	pw := NewProgressWriter(ioutil.Discard)
-	assert.NoError(t, client.UploadFile(fn, buf, pw, true))
-
+	latestRecoveryPointID := "latest-recovery-point-id"
+	gilp := client.getItemLatestPath(latestRecoveryPointID)
+	assert.Equal(t, "/agent/recovery-points/latest-recovery-point-id/path", gilp)
 }
+
+// func TestClient_GetItemLatest(t *testing.T) {
```
Comment (greatbn) [inline pkg/backupapi/file_test.go:27]: If you don't use these code, please remove.

File: pkg/backupapi/volume.go [hunk 2]
```diff
@@ -0,0 +1,192 @@
+package backupapi
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws/awserr"
```
Comment (greatbn) [inline pkg/backupapi/volume.go:6]: group import.

File: pkg/backupapi/volume.go [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+				}
+			}
+		}
+		log.Error("Error PutObject", err)
+		time.Sleep(backoff)
+	}
+	return err
+}
+
+func (c *Client) GetObject(volume volume.StorageVolume, key string, restoreKey *AuthRestore) ([]byte, error) {
```
Comment (greatbn) [inline pkg/backupapi/volume.go:168]: I think we should comment for these exported function. It's doc string.

File: pkg/volume/s3/s3.go [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"math/rand"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/bizflycloud/bizfly-backup/pkg/backupapi"
```
Comment (greatbn) [inline pkg/volume/s3/s3.go:11]: group import.

File: pkg/volume/s3/s3.go [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+		}
+
+		s3.logger = l
+	}
+
+	cred := credentials.NewStaticCredentials(vol.Credential.AwsAccessKeyId, vol.Credential.AwsSecretAccessKey, vol.Credential.Token)
+	//cred := credentials.NewStaticCredentials(accessKey, secretKey, token)
+	_, err := cred.Get()
+	if err != nil {
+		fmt.Println("Bad credentials", err)
```
Comment (greatbn) [inline pkg/volume/s3/s3.go:80]: log or print ?

File: pkg/volume/s3/s3.go [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+		l, err := zap.NewDevelopment()
+		if err != nil {
+			return nil
+		}
+
+		s3.logger = l
+	}
+
+	cred := credentials.NewStaticCredentials(vol.Credential.AwsAccessKeyId, vol.Credential.AwsSecretAccessKey, vol.Credential.Token)
+	//cred := credentials.NewStaticCredentials(accessKey, secretKey, token)
```
Comment (greatbn) [inline pkg/volume/s3/s3.go:77]: do we need this line ?

File: pkg/volume/s3/s3.go [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+			break
+		}
+
+		if aerr, ok := err.(awserr.Error); ok {
+			if aerr.Error() == "Forbidden" {
+				if once {
+					s3.logger.Sugar().Info("Return false cause in put object: ", aerr.Code(), key)
+					return err
+				}
+				s3.logger.Info("====== Put object one more time =============")
```
Comment (greatbn) [inline pkg/volume/s3/s3.go:125]: Why do you use `===`

**Review by greatbn — CHANGES_REQUESTED** [review]
Check my comments

**Review by greatbn — CHANGES_REQUESTED** [review]
Please fix error of CI.

**Review by greatbn — APPROVED** [review]
LGTM

**Pull request merged**